controllers/services_controller.go

```diff
@@ -0,0 +1,144 @@
+package controllers
+
+import (
+	"context"
+	"io/ioutil"
+	"strings"
+
+	"github.com/go-logr/logr"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"kubevirt.io/ssp-operator/internal/common"
+	"kubevirt.io/ssp-operator/internal/operands/metrics"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	MetricsServiceName = "ssp-operator-metrics"
+)
+
+func ServiceObject(namespace string) *v1.Service {
+	policyCluster := v1.ServiceInternalTrafficPolicyCluster
+	familyPolicy := v1.IPFamilyPolicySingleStack
+	return &v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      MetricsServiceName,
+			Namespace: namespace,
+			Labels: map[string]string{
+				"prometheus.ssp.kubevirt.io": "true",
+			},
+			Finalizers: []string{finalizerName},
+		},
+		Spec: v1.ServiceSpec{
+			InternalTrafficPolicy: &policyCluster,
+			IPFamilies:            []v1.IPFamily{v1.IPv4Protocol},
+			IPFamilyPolicy:        &familyPolicy,
+			Ports: []v1.ServicePort{
+				{
+					Name:       metrics.MetricsPortName,
+					Port:       443,
+					Protocol:   v1.ProtocolTCP,
+					TargetPort: intstr.FromString(metrics.MetricsPortName),
+				},
+			},
+			Selector: map[string]string{
+				"prometheus.ssp.kubevirt.io": "true",
+				"name":                       "ssp-operator",
+			},
+			SessionAffinity: v1.ServiceAffinityNone,
+			Type:            v1.ServiceTypeClusterIP,
+		},
+	}
+}
+
+// Annotation to generate RBAC roles to watch and update Pods
+// +kubebuilder:rbac:groups="",resources=services,verbs=get;watch;list;create;update;delete
+
+func CreateServiceController(mgr controllerruntime.Manager) *serviceReconciler {
+	client := mgr.GetClient()
+
+	reconciler := NewServiceReconciler(client)
+
+	return reconciler
+}
+
+func StartServiceController(mgr controllerruntime.Manager, reconciler *serviceReconciler) error {
+	err := reconciler.createMetricsService()
+	if err != nil && !errors.IsAlreadyExists(err) {
+		return err
+	}
+
+	return reconciler.setupController(mgr)
+}
+
+func (r *serviceReconciler) createMetricsService() error {
+	namespace, err := GetOperatorNamespace(r.log)
+	if err != nil {
+		return err
+	}
+
+	service := ServiceObject(namespace)
+
+	return r.client.Create(context.TODO(), service)
+}
+
+func (r *serviceReconciler) setupController(mgr ctrl.Manager) error {
+	predicate, _ := predicate.LabelSelectorPredicate(metav1.LabelSelector{MatchLabels: map[string]string{"prometheus.ssp.kubevirt.io": "true"}})
+	return ctrl.NewControllerManagedBy(mgr).
+		Named("service-controller").
+		For(&v1.Service{}).
```

Comment:
Can we limit the Watch for Services to the namespace we are running in?

Reply:
We are not currently doing this in other places, it is a good idea to check this, let us do it in a separate PR though.